### PR TITLE
basic test code for Accounts

### PR DESCRIPTION
**Pull request by stopthatastronaut**
Adds an OctopusDeployAccount test type. Modelled on the Worker Pool code, but will need some enhancement.

### CODE REVIEW

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    def exists?
+      (!@account.nil?) && (@account != [])
+    end
+
+    def has_description?(account_description)
+      return false if @account.nil?
+      @account["Description"] == account_description
+    end
+
+    def is_account_type?(account_type_name)
```
Comment (matt-richardson) [inline lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: I wonder if some kind of inheritance is the right structure here.
Really depends what the end goal is...

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      return false if @account.nil?
+      @account["AccountType"] == "AzureSubscription"
+    end
+
+    def is_aws_account?
+      return false if @account.nil?
+      @account["AccountType"] == "AmazonWebServicesAccount"
+    end
+
+    def in_space?(space_name)
```
Comment (matt-richardson) [inline lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: not sure this makes sense - already passed in the ctor

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    end
+
+    def is_azure_account?
+      return false if @account.nil?
+      @account["AccountType"] == "AzureSubscription"
+    end
+
+    def is_aws_account?
+      return false if @account.nil?
+      @account["AccountType"] == "AmazonWebServicesAccount"
```
Comment (matt-richardson) [inline lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: maybe use a constant here? and in the methods above?

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      return false if @account.nil?
+      url = "#{@serverUrl}/api/#{@spaceFragment}environments/all?api-key=#{@apiKey}"
+      resp = Net::HTTP.get_response(URI.parse(url))
+      environments = JSON.parse(resp.body)
+      environment_id = environments.select {|e| e["Name"] == environment_name}.first["Id"]
+      !@account["EnvironmentIds"].select {|e| e == environment_id}.empty?
+    end
+  end
+
+  def octopus_deploy_account(serverUrl, apiKey, account_name)
```
Comment (matt-richardson) [inline lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: pass space id

File: spec/octopus/serverspec/type/octopus_deploy_account_spec.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+  context "Server supports spaces" do
+
+    ex_supports_spaces = File.read('spec/octopus/serverspec/json/apisupportsspaces.json')
+    ex_account_found_response = File.read('spec/octopus/serverspec/json/accountsall.json')
+
+    it "handles account found" do
+      stub_request(:get, "https://octopus.example.com/api/").
+          to_return(status: 200, body: ex_supports_spaces, headers: {})
+      stub_request(:get, "https://octopus.example.com/api/Spaces-1/accounts/all?api-key=API-1234567890").
```
Comment (matt-richardson) [inline spec/octopus/serverspec/type/octopus_deploy_account_spec.rb]: same as the other one - might be worth checking against a non default space id

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb (hunk 3 above)
Comment (stopthatastronaut) [reply lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: I was kinda thinking it's a better fit for some kind of enum magic, but my Ruby is not exactly ninja level

like 

```
describe OctopusAccount($url, $apikey, 'my_user_account') do
   it { should be_account_type(OctopusAccount::AwsAccount) }
end

Or something. Actually, a public CONST would achieve much the same effect, wouldn't it?
```

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb (hunk 2 above)
Comment (stopthatastronaut) [reply lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: GORN

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb (hunk 4 above)
Comment (stopthatastronaut) [reply lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: pass space_name, actuellement. Done.

File: spec/octopus/serverspec/type/octopus_deploy_account_spec.rb (hunk 5 above)
Comment (stopthatastronaut) [reply spec/octopus/serverspec/type/octopus_deploy_account_spec.rb]: Good call. Actually, while I'm messing with the example files, I'mma rearrange how they're organised on disk so it represents the API a bit better

File: lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb (hunk 3 above)
Comment (stopthatastronaut) [reply lib/octopus_serverspec_extensions/type/octopus_deploy_account.rb]: Constants have been added. I don't really know if they're ideal. ALso, they probably need to be documented, so I'll update the readme in this PR

File: spec/octopus/serverspec/type/octopus_deploy_account_spec.rb [hunk 6]
```diff
@@ -20,8 +20,8 @@
         to raise_error(/account_name/)
   end
 
-  ex_does_not_support_spaces = File.read('spec/octopus/serverspec/json/apidoesnotsupportspaces.json')
-  ex_acc_found_response = File.read('spec/octopus/serverspec/json/accountsall.json')
+  ex_does_not_support_spaces = get_api_example('/api/2018.4.5')
```
Comment (matt-richardson) [inline spec/octopus/serverspec/type/octopus_deploy_account_spec.rb]: Having `2018.4.5` and `2019.4.5` could cause confusion (the last 2 parts being the same).
Might be better if it was a diff version on one side?

Comment (stopthatastronaut) [reply spec/octopus/serverspec/type/octopus_deploy_account_spec.rb]: Easy enough to change the filename and/or the version

**Pull request merged**